pytorch_lightning/trainer/connectors/callback_connector.py

```diff
@@ -107,3 +108,29 @@ def attach_model_logging_functions(self, model):
         for callback in self.trainer.callbacks:
             callback.log = model.log
             callback.log_dict = model.log_dict
+
+    def attach_model_callbacks(self, model: LightningModule) -> None:
+        """
+        Attaches the callbacks defined in the model.
+        If a callback returned by the model's configure_callback method has the same type as one or several
+        callbacks already present in the trainer callbacks list, it will replace them.
+
+        Args:
+            A model which may or may not define new callbacks in
+            :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_callbacks`.
+        """
+        model_callbacks = model.configure_callbacks()
+        model_callback_types = set(type(c) for c in model_callbacks)
+        trainer_callback_types = set(type(c) for c in self.trainer.callbacks)
+        override_types = model_callback_types.intersection(trainer_callback_types)
+        if override_types:
+            rank_zero_info(
+                "The following callbacks returned in `LightningModule.configure_callbacks` will override"
+                " existing callbacks passed to Trainer:"
+                f" {', '.join(sorted(t.__name__ for t in override_types))}"
+            )
+        # remove all callbacks with a type that occurs in model callbacks
+        all_callbacks = [c for c in self.trainer.callbacks if type(c) not in override_types]
+        all_callbacks.extend(model_callbacks)
+        # TODO: connectors refactor: move callbacks list to connector and do not write Trainer state
+        self.trainer.callbacks = all_callbacks
```

Comment:
i still think we need logic to ensure that any callbacks of type ModelCheckpoint run last. this problem will amplify if we have lightning module callbacks running after the trainer callbacks, though the risk exists today AFAICT. otherwise there's a risk that lightning module state is mutated in another callback after a checkpoint is saved, and that state is silently lost

Reply:
Yes I agree, and I remember it was mentioned several times in the past. Not sure why we haven't done it yet.
Adding a simple mapping here at the end should do it yes?

Reply:
See this [commit ](https://github.com/PyTorchLightning/pytorch-lightning/pull/5621/commits/607ac9fda21e5ed87474dbb41a277cee5b831d36)

Until now it seems we only made sure that the automatic/default checkpoint callback runs last. If the user defines their own callback list, it's not guaranteed. The changes I added will rectify that.

Perhaps I should brake this out into a separate PR

Reply:
> Perhaps I should brake this out into a separate PR

agreed!

Reply:
Won't it be simpler to extend `model_callbacks` with `trainer_callbacks` ?

Reply:
so in the end all callbacks will be attached to the trainer. How the two are merged is simply a matter of definition. We can decide that. 
It could be 
`trainer.callbacks = trainer.callbacks + model_callbacks` or 
`trainer.callbacks = model_callbacks + trainer.callbacks`
both are possible. I don't think one is simpler than the other, we still need to let the callbacks in the model override the ones in trainer (according to william).